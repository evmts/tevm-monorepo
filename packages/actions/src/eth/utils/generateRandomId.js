/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
