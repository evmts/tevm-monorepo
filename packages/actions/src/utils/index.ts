export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
