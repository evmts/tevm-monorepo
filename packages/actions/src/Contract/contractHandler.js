import { createAddress } from '@tevm/address'
import { DecodeFunctionDataError, InvalidRequestError, RevertError } from '@tevm/errors'
import { decodeErrorResult, decodeFunctionResult, encodeFunctionData, isHex } from '@tevm/utils'
import { callHandler } from '../Call/callHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateContractParams } from './validateContractParams.js'

/**
 * Creates a tree-shakable instance of `contractHandler` for handling contract interactions with the Ethereumjs EVM.
 * This function uses `callHandler` under the hood to execute contract calls.
 *
 * Note: This is the internal logic used by higher-level APIs such as `tevmContract`.
 *
 * @param {import('@tevm/node').TevmNode} client - The TEVM base client instance.
 * @param {object} [options] - Optional parameters.
 * @param {boolean} [options.throwOnFail=true] - Whether to throw an error on failure.
 * @returns {import("../Contract/ContractHandlerType.js").ContractHandler} The contract handler function.
 * @throws {import('./TevmCallError.js').TevmCallError} If `throwOnFail` is true, returns `TevmCallError` as value.
 *
 * @example
 * ```typescript
 * import { createTevmNode } from 'tevm/node'
 * import { contractHandler } from 'tevm/actions'
 *
 * const client = createTevmNode()
 *
 * const contract = contractHandler(client)
 *
 * const res = await contract({
 *   to: `0x${'69'.repeat(20)}`,
 *   abi: [{...}], // ABI array
 *   functionName: 'myFunction',
 *   args: [1, 2, 3],
 * })
 * ```
 */
export const contractHandler =
	(client, { throwOnFail: throwOnFailDefault = true } = {}) =>
	async (params) => {
		const errors = validateContractParams(/** @type any*/ (params))
		if (errors.length > 0) {
			client.logger.debug(errors, 'contractHandler: Invalid params')
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				errors,
				executionGasUsed: 0n,
				rawData: '0x',
			})
		}
		const vm = await client.getVm()

		const contract = params.to && (await vm.evm.stateManager.getContractCode(createAddress(params.to)))
		const precompile = params.to && vm.evm.getPrecompile(createAddress(params.to))
		if (!params.deployedBytecode && !params.code && contract && contract?.length === 0 && !precompile) {
			client.logger.debug(
				{ contract, precompile, to: params.to },
				'contractHandler: No contract bytecode nor precompile was found at specified `to` address. Unable to execute contract call.',
			)
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [
					{
						_tag: 'InvalidRequestError',
						name: 'InvalidRequestError',
						message: `Contract at address ${params.to} does not exist`,
					},
				],
			})
		}

		let functionData
		try {
			functionData = encodeFunctionData(
				/** @type {any} */ ({
					abi: params.abi,
					functionName: params.functionName,
					args: params.args,
				}),
			)
		} catch (e) {
			client.logger.debug(e, 'contractHandler: Unable to encode the abi, functionName, and args into hex data')
			const cause = /** @type {Error}*/ (e)
			const err = new InvalidRequestError(cause.message, { cause })
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [err],
			})
		}

		client.logger.debug(
			functionData,
			'contractHandler: Encoded data, functionName, and args into hex data to execute call',
		)

		const result = await callHandler(client, {
			throwOnFail: throwOnFailDefault,
		})({
			...params,
			throwOnFail: false,
			data: functionData,
		})

		if (result.errors && result.errors.length > 0) {
			result.errors = result.errors.map((err) => {
				if (isHex(err.message) && err instanceof RevertError) {
					client.logger.debug(err, 'contractHandler: Contract revert error. Decoding the error')
					/**
					 * @type {undefined |ReturnType<typeof decodeErrorResult>}
					 */
					let decodedError = undefined
					try {
						decodedError = decodeErrorResult(
							/** @type {any} */ ({
								abi: params.abi,
								data: err.message,
								functionName: params.functionName,
							}),
						)
						const message = `Revert: ${decodedError?.errorName ?? `There was a revert with no revert message ${err.message}`}`
						client.logger.debug(err, message)
						return new RevertError(message, { cause: err })
					} catch (e) {
						client.logger.debug(e, 'There was an error decoding error result')
						return err
					}
				}
				return err
			})
			client.logger.debug(result.errors, 'contractHandler: Execution errors')
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
		}

		let decodedResult
		try {
			decodedResult = decodeFunctionResult(
				/** @type {any} */ ({
					abi: params.abi,
					data: result.rawData,
					functionName: params.functionName,
				}),
			)
		} catch (e) {
			if (result.rawData === '0x') {
				throw new Error(
					'0x data returned from EVM with no error message. This indicates a the contract was missing or bug in Tevm if no other errors were thrown',
				)
			}
			client.logger.debug(e, 'contractHandler: Error decoding returned call data with provided abi and functionName')
			const cause = /** @type {Error}*/ (e)
			const err = new DecodeFunctionDataError(cause.message, { cause })
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				debugContext: {
					abi: params.abi,
					rawData: result.rawData,
					functionName: params.functionName,
				},
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [err],
			})
		}

		client.logger.debug(decodedResult, 'contractHandler: decoded data into a final result')

		return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
			.../** @type any */ (result),
			data: decodedResult,
		})
	}
