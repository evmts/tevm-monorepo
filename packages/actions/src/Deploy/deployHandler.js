import { InvalidRequestError, RevertError } from '@tevm/errors'
import { encodeDeployData } from '@tevm/utils'
import { decodeErrorResult, isHex } from '@tevm/utils'
import { callHandler } from '../Call/callHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'

/**
 * Creates a tree-shakable instance of `deployHandler` for handling the deployment of contracts to TEVM.
 * This function uses `callHandler` under the hood to execute the deployment.
 *
 * Note: This is the internal logic used by higher-level APIs such as `tevmDeploy`.
 *
 * @param {import('@tevm/node').TevmNode} client - The TEVM base client instance.
 * @param {object} [options] - Optional parameters.
 * @param {boolean} [options.throwOnFail=true] - Whether to throw an error on failure.
 * @returns {import("../Deploy/DeployHandlerType.js").DeployHandler} The deploy handler function.
 * @throws {import('./TevmCallError.js').TevmCallError} If `throwOnFail` is true, returns `TevmCallError` as value.
 *
 * @example
 * ```typescript
 * import { createTevmNode } from 'tevm/node'
 * import { deployHandler } from 'tevm/actions'
 *
 * const client = createTevmNode()
 *
 * const deploy = deployHandler(client)
 *
 * const res = await deploy({
 *   bytecode: '0x...', // Contract bytecode
 *   abi: [{...}], // ABI array
 *   args: [1, 2, 3], // Constructor arguments
 *   createTransaction: true,
 * })
 * ```
 */
export const deployHandler =
	(client, { throwOnFail: throwOnFailDefault = true } = {}) =>
	async (params) => {
		client.logger.debug(params, 'deployHandler called with params')
		// TODO Moving fast atm we should ideally add validators to zod
		let deployData
		try {
			deployData = encodeDeployData(
				/** @type {any} */ ({
					abi: params.abi,
					bytecode: params.bytecode,
					args: params.args,
				}),
			)
		} catch (e) {
			client.logger.debug(e, 'deployHandler: Unable to encode the abi, functionName, and args into hex data')
			const cause = /** @type {Error}*/ (e)
			const err = new InvalidRequestError(cause.message, { cause })

			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				/**
				 * @type {import('@tevm/utils').Hex}
				 */
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [err],
			})
		}

		client.logger.debug(deployData, 'deployHandler: Encoded abi bytecode and args into hex data to execute call')

		const result = await callHandler(client, {
			throwOnFail: throwOnFailDefault,
		})({
			...params,
			// Unlike most calls deployments default to true
			createTransaction: params.createTransaction !== undefined ? params.createTransaction : true,
			data: deployData,
			throwOnFail: false,
		})

		if (result.errors && result.errors.length > 0) {
			result.errors = result.errors.map((err) => {
				if (isHex(err.message) && err instanceof RevertError) {
					try {
						client.logger.debug(err, 'contractHandler: Contract revert error. Decoding the error')
						const decodedError = decodeErrorResult(
							/** @type {any} */ ({
								abi: params.abi,
								data: err.message,
								functionName: 'constructor',
							}),
						)
						const message = `Revert: ${decodedError.errorName} ${JSON.stringify(decodedError)}`
						client.logger.debug(message, 'Revert message decoded')
						return new RevertError(message, { cause: /** @type any*/ (err) })
					} catch (e) {
						client.logger.warn(e, 'Unable to decode revert data')
						return err
					}
				}
				return err
			})
			client.logger.debug(result.errors, 'contractHandler: Execution errors')
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
		}

		return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
	}
