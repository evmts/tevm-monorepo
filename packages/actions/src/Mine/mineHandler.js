import { InternalError, MisconfiguredClientError, UnreachableCodeError } from '@tevm/errors'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { emitEvents } from './emitEvents.js'
import { validateMineParams } from './validateMineParams.js'

// TODO Errors can leave us in bad states

/**
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./MineHandlerType.js').MineHandler}
 */
export const mineHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params } = {}) => {
		client.logger.debug({ throwOnFail, ...params }, 'mineHandler called with params')
		const errors = validateMineParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}
		const { interval = 1, blockCount = 1 } = params

		/**
		 * @type {Array<import('@tevm/block').Block>}
		 */
		const newBlocks = []
		/**
		 * @type {Map<import('@tevm/utils').Hex,Array<import('@tevm/receipt-manager').TxReceipt>>}
		 */
		const newReceipts = new Map()

		client.logger.debug({ blockCount }, 'processing txs')
		const pool = await client.getTxPool()
		const originalVm = await client.getVm()

		switch (client.status) {
			case 'MINING': {
				// wait for the previous mine to finish
				await new Promise((resolve) => {
					client.on('newBlock', async () => {
						if (client.status === 'MINING') {
							return
						}
						client.status = 'MINING'
						resolve(client)
					})
				})
				break
			}
			case 'INITIALIZING': {
				await client.ready()
				client.status = 'MINING'
				break
			}
			case 'SYNCING': {
				const err = new MisconfiguredClientError('Syncing not currently implemented')
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
			case 'STOPPED': {
				const err = new MisconfiguredClientError('Client is stopped')
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
			case 'READY': {
				client.status = 'MINING'
				break
			}
			default: {
				const err = new UnreachableCodeError(client.status)
				return maybeThrowOnFail(throwOnFail, { errors: [err] })
			}
		}

		const vm = await originalVm.deepCopy()
		const receiptsManager = await client.getReceiptsManager()

		for (let count = 0; count < blockCount; count++) {
			const parentBlock = await vm.blockchain.getCanonicalHeadBlock()

			let timestamp = Math.max(Math.floor(Date.now() / 1000), Number(parentBlock.header.timestamp))
			timestamp = count === 0 ? timestamp : timestamp + interval

			const blockBuilder = await vm.buildBlock({
				parentBlock,
				headerData: {
					timestamp,
					number: parentBlock.header.number + 1n,
					// The following 2 are currently not supported
					// difficulty: undefined,
					// coinbase,
					gasLimit: parentBlock.header.gasLimit,
					baseFeePerGas: parentBlock.header.calcNextBaseFee(),
				},
				blockOpts: {
					// Proof of authority not currently supported
					// cliqueSigner,
					// proof of work not currently supported
					//calcDifficultyFromHeader,
					//ck
					freeze: false,
					setHardfork: false,
					putBlockIntoBlockchain: false,
					common: vm.common,
				},
			})
			// TODO create a Log manager
			const orderedTx = await pool.txsByPriceAndNonce({ baseFee: parentBlock.header.calcNextBaseFee() })

			let index = 0
			// TODO we need to actually handle this
			const blockFull = false
			/**
			 * @type {Array<import('@tevm/receipt-manager').TxReceipt>}
			 */
			const receipts = []
			while (index < orderedTx.length && !blockFull) {
				const nextTx = /** @type {import('@tevm/tx').TypedTransaction}*/ (orderedTx[index])
				client.logger.debug(bytesToHex(nextTx.hash()), 'new tx added')
				const txResult = await blockBuilder.addTransaction(nextTx, {
					skipHardForkValidation: true,
				})
				if (txResult.execResult.exceptionError) {
					if (txResult.execResult.exceptionError.error === 'out of gas') {
						client.logger.debug(txResult.execResult.executionGasUsed, 'out of gas')
					}
					client.logger.debug(
						txResult.execResult.exceptionError,
						`There was an exception when building block for tx ${bytesToHex(nextTx.hash())}`,
					)
				}
				receipts.push(txResult.receipt)
				index++
			}
			await vm.stateManager.checkpoint()
			const createNewStateRoot = true
			await vm.stateManager.commit(createNewStateRoot)
			const block = await blockBuilder.build()
			await Promise.all([receiptsManager.saveReceipts(block, receipts), vm.blockchain.putBlock(block)])
			pool.removeNewBlockTxs([block])

			newBlocks.push(block)
			newReceipts.set(bytesToHex(block.hash()), receipts)

			const value = vm.stateManager._baseState.stateRoots.get(bytesToHex(block.header.stateRoot))

			if (!value) {
				throw new InternalError(
					'InternalError: State root not found in mineHandler. This indicates a potential inconsistency in state management.',
				)
			}

			originalVm.stateManager.saveStateRoot(block.header.stateRoot, value)
		}
		originalVm.blockchain = vm.blockchain
		originalVm.evm.blockchain = vm.evm.blockchain
		await originalVm.stateManager.setStateRoot(hexToBytes(vm.stateManager._baseState.getCurrentStateRoot()))

		client.status = 'READY'

		emitEvents(client, newBlocks, newReceipts)

		return { blockHashes: newBlocks.map((b) => bytesToHex(b.hash())) }
	}
