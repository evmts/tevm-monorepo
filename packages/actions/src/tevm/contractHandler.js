import { callHandler } from './callHandler.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAddress } from '@tevm/utils'
import {
	decodeErrorResult,
	decodeFunctionResult,
	encodeFunctionData,
	isHex,
} from '@tevm/utils'
import { validateContractParams } from '@tevm/zod'

/**
 * Creates an ContractHandler for handling contract params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'getVm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import("@tevm/actions-types").ContractHandler}
 */
export const contractHandler =
	(client, { throwOnFail: throwOnFailDefault = true } = {}) =>
	async (params) => {
		const errors = validateContractParams(/** @type any*/ (params))
		if (errors.length > 0) {
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				errors,
				executionGasUsed: 0n,
				rawData: '0x',
			})
		}

		const vm = await client.getVm()

		const contract = await vm.evm.stateManager.getContractCode(
			EthjsAddress.fromString(params.to),
		)
		const precompile =
			params.to && vm.evm.getPrecompile(EthjsAddress.fromString(params.to))
		if (contract.length === 0 && !precompile) {
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [
					{
						_tag: 'InvalidRequestError',
						name: 'InvalidRequestError',
						message: `Contract at address ${params.to} does not exist`,
					},
				],
			})
		}

		let functionData
		try {
			functionData = encodeFunctionData(
				/** @type {any} */ ({
					abi: params.abi,
					functionName: params.functionName,
					args: params.args,
				}),
			)
		} catch (e) {
			/**
			 * @type {import('@tevm/errors').InvalidRequestError}
			 */
			const err = {
				name: 'InvalidRequestError',
				_tag: 'InvalidRequestError',
				message: /** @type {Error}*/ (e).message,
			}
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [err],
			})
		}

		const result = await callHandler(client, {
			throwOnFail: throwOnFailDefault,
		})({
			...params,
			throwOnFail: false,
			data: functionData,
		})

		if (result.errors && result.errors.length > 0) {
			result.errors = result.errors.map((err) => {
				if (isHex(err.message) && err._tag === 'revert') {
					const decodedError = decodeErrorResult(
						/** @type {any} */ ({
							abi: params.abi,
							data: err.message,
							functionName: params.functionName,
						}),
					)
					return {
						...err,
						message: `Revert: ${decodedError.errorName} ${JSON.stringify(
							decodedError,
						)}`,
					}
				}
				return err
			})
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
		}

		let decodedResult
		try {
			decodedResult = decodeFunctionResult(
				/** @type {any} */ ({
					abi: params.abi,
					data: result.rawData,
					functionName: params.functionName,
				}),
			)
		} catch (e) {
			/**
			 * @type {import('@tevm/errors').ContractError}
			 */
			const err = {
				name: 'DecodeFunctionDataError',
				_tag: 'DecodeFunctionDataError',
				message: /** @type {Error}*/ (e).message,
			}
			return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
				rawData: '0x',
				executionGasUsed: 0n,
				errors: [err],
			})
		}

		return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
			.../** @type any */ (result),
			data: decodedResult,
		})
	}
