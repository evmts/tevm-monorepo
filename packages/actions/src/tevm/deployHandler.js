import { callHandler } from './callHandler.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { encodeDeployData } from '@tevm/utils'
import {
  decodeErrorResult,
  isHex,
} from '@tevm/utils'

// TODO let's add warnings to other calls such that if the contract is still in mempool we throw a warning telling user they need to mine a block

/**
 * Creates an DeployHandler for handling deploying a contract to tevm
 * @param {import('@tevm/base-client').BaseClient} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import("@tevm/actions-types").DeployHandler}
 */
export const deployHandler =
  (client, { throwOnFail: throwOnFailDefault = true } = {}) =>
    async (params) => {
      client.logger.debug(params, 'deployHandler called with params')
      // TODO Moving fast atm we should ideally add validators to @tevm/zod
      let deployData
      try {
        deployData = encodeDeployData(
				/** @type {any} */({
            abi: params.abi,
            bytecode: params.bytecode,
            args: params.args,
          }),
        )
      } catch (e) {
        client.logger.debug(
          e,
          'deployHandler: Unable to encode the abi, functionName, and args into hex data',
        )
        /**
         * @type {import('@tevm/errors').InvalidRequestError}
         */
        const err = {
          name: 'InvalidRequestError',
          _tag: 'InvalidRequestError',
          message: /** @type {Error}*/ (e).message,
        }
        return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, {
          /**
           * @type {import('@tevm/utils').Hex}
           */
          rawData: '0x',
          executionGasUsed: 0n,
          errors: [err],
        })
      }

      client.logger.debug(
        deployData,
        'deployHandler: Encoded abi bytecode and args into hex data to execute call',
      )

      const result = await callHandler(client, {
        throwOnFail: throwOnFailDefault,
      })({
        ...params,
        // Unlike most calls deployments default to true
        createTransaction: params.createTransaction !== undefined ? params.createTransaction : true,
        data: deployData,
        throwOnFail: false,
      })

      if (result.errors && result.errors.length > 0) {
        result.errors = result.errors.map((err) => {
          if (isHex(err.message) && err._tag === 'revert') {
            client.logger.debug(
              err,
              'contractHandler: Contract revert error. Decoding the error',
            )
            const decodedError = decodeErrorResult(
						/** @type {any} */({
                abi: params.abi,
                data: err.message,
                functionName: 'constructor',
              }),
            )
            const message = `Revert: ${decodedError.errorName} ${JSON.stringify(
              decodedError,
            )}`
            client.logger.debug(message, 'Revert message decoded')
            return {
              ...err,
              message,
            }
          }
          return err
        })
        client.logger.debug(result.errors, 'contractHandler: Execution errors')
        return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
      }

      return maybeThrowOnFail(params.throwOnFail ?? throwOnFailDefault, result)
    }
