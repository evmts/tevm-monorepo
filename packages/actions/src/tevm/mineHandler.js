import { bytesToHex } from '@tevm/utils'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { validateMineParams } from '@tevm/zod'

// TODO Errors can leave us in bad states

/**
 * @param {import("@tevm/base-client").BaseClient} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').MineHandler}
 */
export const mineHandler =
  (client, options = {}) =>
    async ({ throwOnFail = options.throwOnFail ?? true, ...params } = {}) => {
      client.logger.debug({ throwOnFail, ...params }, 'mineHandler called with params')
      const errors = validateMineParams(params)
      if (errors.length > 0) {
        return maybeThrowOnFail(throwOnFail, { errors })
      }
      const { interval = 1, blockCount = 1 } = params

      /**
       * @type {Array<import('@tevm/utils').Hex>}
       */
      const blockHashes = []

      client.logger.debug({ blockCount }, 'processing txs')

      for (let count = 0; count < blockCount; count++) {
        const pool = await client.getTxPool()
        const vm = await client.getVm()
        const receiptsManager = await client.getReceiptsManager()
        const parentBlock = await vm.blockchain.getCanonicalHeadBlock()

        let timestamp = Math.max(Math.floor(Date.now() / 1000), Number(parentBlock.header.timestamp))
        timestamp = count === 0 ? timestamp : timestamp + interval

        const blockBuilder = await vm.buildBlock({
          parentBlock,
          headerData: {
            timestamp,
            number: parentBlock.header.number + 1n,
            // The following 2 are currently not supported
            // difficulty: undefined,
            // coinbase,
            gasLimit: parentBlock.header.gasLimit,
            baseFeePerGas: parentBlock.header.calcNextBaseFee(),
          },
          blockOpts: {
            // Proof of authority not currently supported
            // cliqueSigner,
            // proof of work not currently supported
            //calcDifficultyFromHeader,
            //ck
            freeze: false,
            setHardfork: false,
            putBlockIntoBlockchain: false,
            common: vm.common
          },
        })
        // TODO create a Log manager
        const orderedTx = await pool.txsByPriceAndNonce({ baseFee: parentBlock.header.calcNextBaseFee() })

        let index = 0
        let blockFull = false
        /**
         * @type {Array<import('@tevm/receipt-manager').TxReceipt>}
         */
        const receipts = []
        while (index < orderedTx.length && !blockFull) {
          const nextTx = /** @type {import('@tevm/tx').TypedTransaction}*/(orderedTx[index])
          client.logger.debug(bytesToHex(nextTx.hash()), 'new tx added')
          const txResult = await blockBuilder.addTransaction(nextTx, {
            skipHardForkValidation: true,
          })
          if (txResult.execResult.exceptionError) {
            if (txResult.execResult.exceptionError.error === 'out of gas') {
              client.logger.debug(txResult.execResult.executionGasUsed, 'out of gas')
            }
            client.logger.debug(txResult.execResult.exceptionError, `There was an exception when building block for tx ${bytesToHex(nextTx.hash())}`)
          }
          receipts.push(txResult.receipt)
          index++
        }
        await vm.stateManager.checkpoint()
        const createNewStateRoot = true
        await vm.stateManager.commit(createNewStateRoot)
        const block = await blockBuilder.build()
        await Promise.all([
          receiptsManager.saveReceipts(block, receipts),
          vm.blockchain.putBlock(block),
        ])
        pool.removeNewBlockTxs([block])

        blockHashes.push(bytesToHex(block.hash()))
      }
      return { blockHashes }
    }
