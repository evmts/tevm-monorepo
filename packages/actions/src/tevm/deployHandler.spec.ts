import { createBaseClient } from '@tevm/base-client'
import { describe, expect, it } from 'bun:test'
import { deployHandler } from './deployHandler.js'
import { contractHandler } from './contractHandler.js'
import { EthjsAddress, bytesToHex, parseAbi, type Hex } from '@tevm/utils'
import { mineHandler } from './mineHandler.js'

// This contract exists as SimpleConstructor in bench package. Consider moving it to test utils
const simpleConstructorBytecode = '0x608060405234801561000f575f80fd5b506040516101ef3803806101ef83398181016040528101906100319190610074565b805f819055505061009f565b5f80fd5b5f819050919050565b61005381610041565b811461005d575f80fd5b50565b5f8151905061006e8161004a565b92915050565b5f602082840312156100895761008861003d565b5b5f61009684828501610060565b91505092915050565b610143806100ac5f395ff3fe608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c63430008160033' as const
const simpleConstructorAbi = parseAbi([
  'constructor(uint256 _initialValue)',
  'function get() view returns (uint256)',
  'function set(uint256 x)'
])
describe('deployHandler', () => {
  it('should deploy a contract', async () => {
    const client = createBaseClient({ loggingLevel: 'warn' })
    const initialValue = 420n
    // TODO this type is fucked atm
    const deployResult = await deployHandler(client)({
      abi: simpleConstructorAbi,
      bytecode: simpleConstructorBytecode,
      args: [initialValue]
    })
    expect(
      deployResult
    ).toEqual({
      "createdAddress": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
      "createdAddresses":
        new Set([
          "0x5FbDB2315678afecb367f032d93F642f64180aa3",
        ]),
      "executionGasUsed": 87131n,
      "gas": 16754684n,
      "logs": [],
      "rawData": "0x608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c63430008160033",
      "selfdestruct": new Set(),
      "txHash": "0x2a872fc2c05d90cbbdfbed7a5c831533dc1d02c1be4ab374b7d9c66e9ccec0e8",

    })
    const mineResult = await mineHandler(client)()
    const vm = await client.getVm()
    expect(mineResult.errors).toBeUndefined()
    expect(mineResult.blockHashes).toHaveLength(1)

    // TODO this is broke
    // expect(await ethGetTransactionReceiptHandler(client)({ hash: deployResult.txHash as Hex })).toEqual({})

    expect(bytesToHex(await vm.stateManager.getContractCode(EthjsAddress.fromString(deployResult.createdAddress as Hex)))).toEqual('0x608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c63430008160033')
    // TODO this is broke too
    // expect((await vm.stateManager.getAccount(EthjsAddress.fromString(deployResult.createdAddress as Hex)))?.isContract()).toBeTrue()
    expect(
      (await contractHandler(client)({
        abi: simpleConstructorAbi,
        functionName: 'get',
        to: deployResult.createdAddress as Hex
      })).data
    ).toBe(initialValue)
  })
})
