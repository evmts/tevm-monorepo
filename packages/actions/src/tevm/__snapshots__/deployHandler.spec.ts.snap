// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`deployHandler should deploy a contract 1`] = `
{
  "createdAddress": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
  "createdAddresses":   
Set {
    "0x5FbDB2315678afecb367f032d93F642f64180aa3",
  }
,
  "executionGasUsed": 87131n,
  "gas": 16754684n,
  "logs": [],
  "rawData": "0x608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c63430008160033",
  "selfdestruct":   Set {},
  "txHash": "0xcba00d9e62b11567e58af0ed041c70abe437e9c1c621043a148546e0ff1abad9",
}
`;
